language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,615 @@
+        env = PLC.TypeCheckEnv config mempty mempty
+
+-- | Extend the context of a 'TypeCheckM' computation with a kinded variable.
+withTyVar :: TyName -> Kind () -> TypeCheckM uni ann a -> TypeCheckM uni ann a
+withTyVar name = local . over PLC.tceTyVarKinds . insertByName name
+
+-- | Extend the context of a 'TypeCheckM' computation with a typed variable.
+withVar :: Name -> Normalized (Type TyName uni ()) -> TypeCheckM uni ann a -> TypeCheckM uni ann a
+withVar name = local . over PLC.tceVarTypes . insertByName name . pure
+
+-- | Look up a 'DynamicBuiltinName' in the 'DynBuiltinNameTypes' environment.
+lookupDynamicBuiltinNameM
+    :: ann -> PLC.DynamicBuiltinName -> TypeCheckM uni ann (Normalized (Type TyName uni ()))
+lookupDynamicBuiltinNameM ann name = do
+    PLC.DynamicBuiltinNameTypes dbnts <- asks $ PLC._tccDynamicBuiltinNameTypes . PLC._tceTypeCheckConfig
+    case Map.lookup name dbnts of
+        Nothing ->
+            throwError $ UnknownDynamicBuiltinName ann (UnknownDynamicBuiltinNameErrorE name)
+        Just ty -> liftDupable ty
+
+-- | Look up a type variable in the current context.
+lookupTyVarM :: ann -> TyName -> TypeCheckM uni ann (Kind ())
+lookupTyVarM ann name = do
+    mayKind <- asks $ lookupName name . PLC._tceTyVarKinds
+    case mayKind of
+        Nothing   -> throwError $ FreeTypeVariableE ann name
+        Just kind -> pure kind
+
+-- | Look up a term variable in the current context.
+lookupVarM :: ann -> Name -> TypeCheckM uni ann (Normalized (Type TyName uni ()))
+lookupVarM ann name = do
+    mayTy <- asks $ lookupName name . PLC._tceVarTypes
+    case mayTy of
+        Nothing -> throwError $ FreeVariableE ann name
+        Just ty -> liftDupable ty
+
+-- #############
+-- ## Dummies ##
+-- #############
+
+dummyUnique :: Unique
+dummyUnique = Unique 0
+
+dummyTyName :: TyName
+dummyTyName = TyName (Name "*" dummyUnique)
+
+dummyKind :: Kind ()
+dummyKind = Type ()
+
+dummyType :: Type TyName uni ()
+dummyType = TyVar () dummyTyName
+
+dummyTerm :: Term tyname Name unia ()
+dummyTerm = Var () (Name "__var" dummyUnique)
+
+-- ########################
+-- ## Type normalization ##
+-- ########################
+
+-- | Normalize a 'Type'.
+normalizeTypeM :: Type TyName uni ann1 -> TypeCheckM uni ann (Normalized (Type TyName uni ann1))
+normalizeTypeM ty = Norm.runNormalizeTypeM $ Norm.normalizeTypeM ty
+
+-- | Substitute a type for a variable in a type and normalize the result.
+substNormalizeTypeM
+    :: Normalized (Type TyName uni ())  -- ^ @ty@
+    -> TyName                           -- ^ @name@
+    -> Type TyName uni ()               -- ^ @body@
+    -> TypeCheckM uni ann (Normalized (Type TyName uni ()))
+substNormalizeTypeM ty name body = Norm.runNormalizeTypeM $ Norm.substNormalizeTypeM ty name body
+
+-- ###################
+-- ## Kind checking ##
+-- ###################
+
+-- | Infer the kind of a type.
+inferKindM :: Type TyName uni ann -> TypeCheckM uni ann (Kind ())
```

Comment:
It feels like we should be able to reuse the kind-checking machinery from PLC, after all that is totally the same. I guess the reason why we can't is that they have different monads? And the only reason they have separate monads is because they have different errors? 

I think we could fix this if we wanted, by:
- Making the PLC typechecker functions use `MonadError` etc rather than a concrete monad. Then we can instantiate them for PLC and PIR.
- Making the PLC type errors a subset of the PIR type errors. Then we can use the prismatic error handling to put `AsTypeError` on the PLC functions and have them work for the PIR type error type too.

Not sure if this would be worth it, but I'd be inclined to try it.

Reply:
This would also let us reuse a bunch of the utility functions that are currently duplicated.

Reply:
Actually, there's a simpler approach: we can probably write a function `PLC.TypecheckM uni ann a -> PIR.TypecheckM uni ann a`, which would do the job.